bentoml/yatai/grpc_stream_utils.py

```diff
@@ -0,0 +1,138 @@
+import os
+import io
+import tarfile
+from bentoml.yatai.proto.repository_pb2 import UploadBentoRequest, DownloadBentoResponse
+
+
+DEFAULT_FILE_CHUNK_SIZE = 1024 * 8
+
+
+class FileStream:
+    def __init__(self):
+        self.buffer = io.BytesIO()
+        self.offset = 0
+
+    def write(self, s):
+        self.buffer.write(s)
+        self.offset += len(s)
+
+    def tell(self):
+        return self.offset
+
+    def close(self):
+        self.buffer.close()
+
+    def read_value(self):
+        try:
+            return self.buffer.getvalue()
+        finally:
+            self.buffer.close()
+            self.buffer = io.BytesIO()
+
+
+class BentoBundleStreamRequestsOrResponses:
+    def __init__(
+        self,
+        bento_name,
+        bento_version,
+        directory_path,
+        file_chunk_size=DEFAULT_FILE_CHUNK_SIZE,
+        is_request=True,
+    ):
+        """
+        Build a tar file chunk by chunk on the fly from a directory and then stream out
+        either request or response for upload/download.
+
+        Args:
+            bento_name:
+            bento_version:
+            directory_path:
+            file_chunk_size:
+            is_request:
+        """
+        self.bento_name = bento_name
+        self.bento_version = bento_version
+        self.directory_path = directory_path
+        self.file_chunk_size = file_chunk_size
+        self.is_request = is_request
```

Comment:
might be better to just separate them to two different classes and have them use some shared utility functions?